### PR TITLE
[FX][export] strict DCE pass, check schema for node impurity

### PR DESCRIPTION
**Pull request by yushangdi**
Fixes the failure in `test/export/test_export_training_ir_to_run_decomp.py ` caused by dead code elimination removing node with side effects.

For background, in export, we may want to export higher-level IRs that are not functional, so we need to check for side effects more carefully.

 A call_function node is impure if it has at least one mutable argument.

Fixed the tests below:

test_to_module_with_mutated_buffer_multiple_update_sub_later
test_export_input_mutation_static_shape
test_buffer_util

Another attempt modifying the original DCE pass is made in PR #130395, but it breaks some other tests, so here we add a flag and use it for export only.

### CODE REVIEW

**Comment by pianpwk** [comment]
Since this is basically adding a custom version of DCE that only export is currently using, what do you think about just having  something like an `is_impure_fn` argument to the DCE call? The schema checking logic can be moved to the export side, and if the desirable DCE behavior changes for us, or for some other user, it's still easy to extend. But your decision :)

Thanks for the change!

**Review by pianpwk — APPROVED** [review]
You can resolve the backward compatibility issues by adding default arguments here: https://github.com/pytorch/pytorch/blob/main/test/expect/TestFXAPIBackwardCompatibility.test_function_back_compat-fx_backcompat_function_signatures.expect

I added a comment, but overall looks good.

**Comment by yushangdi** [comment]
> Since this is basically adding a custom version of DCE that only export is currently using, what do you think about just having something like an `is_impure_fn` argument to the DCE call? The schema checking logic can be moved to the export side, and if the desirable DCE behavior changes for us, or for some other user, it's still easy to extend. But your decision :)
> 
> Thanks for the change!

I think that's a good idea, thanks for the suggestion! I made the change.

File: torch/fx/graph.py [hunk 1]
```diff
@@ -1625,6 +1625,34 @@ def forward(self, x):
 
         return changed
 
+    def _eliminate_dead_code_custom(self, is_impure_node: Callable[[Node], bool]):
```
Comment (pianpwk) [inline torch/fx/graph.py:1628]: I think this can just be a signature change to `eliminate_dead_code()`?

Comment (yushangdi) [reply torch/fx/graph.py]: fixed now!

Comment (pianpwk) [reply torch/fx/graph.py]: nice! this looks great

**Pull request closed without merging**